File: src/sync/websocket_client.py
"""
WebSocket client for the ICC multi-device sync system (desktop side).

Connects to other desktop peers or relay servers via WebSocket,
handles authentication, and relays clipboard sync messages.

Binary frame protocol (consistent with ProtocolCodec):
    [2 bytes: header length (big-endian uint16)]
    [N bytes: header JSON (UTF-8)]
    [M bytes: binary payload (optional)]
"""

from __future__ import annotations

import asyncio
import logging
import time
from typing import Any, Callable, Dict, Optional, Set, Tuple

import websockets
from websockets.client import WebSocketClientProtocol

from .protocol import (
    MSG_TYPE_AUTH_CHALLENGE,
    MSG_TYPE_AUTH_FAIL,
    MSG_TYPE_AUTH_OK,
    MSG_TYPE_CLIPBOARD_SYNC,
    MSG_TYPE_CLIPBOARD_SYNC_BINARY,
    MSG_TYPE_DEVICE_LIST,
    MSG_TYPE_DEVICE_OFFLINE,
    MSG_TYPE_DEVICE_ONLINE,
    MSG_TYPE_HELLO,
    MSG_TYPE_PAIRING_CONFIRM,
    MSG_TYPE_PAIRING_REQUEST,
    MSG_TYPE_PING,
    MSG_TYPE_PONG,
    ProtocolCodec,
    SyncMessage,
)
from .auth import compute_hmac, generate_challenge, verify_hmac
from .device_manager import DeviceInfo, DeviceManager

logger = logging.getLogger(__name__)


class WebSocketClient:
    """WebSocket client for connecting to sync peers or relay servers.

    Manages:
    - Outgoing WebSocket connections with automatic reconnection
    - HMAC-SHA256 authentication handshake (client side)
    - Message sending and receiving
    - Heartbeat (ping/pong) responses
    """

    def __init__(
        self,
        device_manager: Optional[DeviceManager] = None,
        secret_store: Optional[Dict[str, str]] = None,
        on_clipboard_sync: Optional[Callable] = None,
        on_device_online: Optional[Callable] = None,
        on_device_offline: Optional[Callable] = None,
        reconnect_interval: float = 5.0,
    ) -> None:
        """Initialize the WebSocket client.

        Args:
            device_manager: DeviceManager for looking up paired devices.
            secret_store: Dictionary mapping device IDs to shared secrets.
            on_clipboard_sync: Callback for clipboard sync messages.
            on_device_online: Callback when a remote device comes online.
            on_device_offline: Callback when a remote device goes offline.
            reconnect_interval: Base interval in seconds between reconnection attempts.
        """
        self.device_manager = device_manager or DeviceManager()
        self._secret_store = secret_store or {}
        self.on_clipboard_sync = on_clipboard_sync
        self.on_device_online = on_device_online
        self.on_device_offline = on_device_offline
        self.reconnect_interval = reconnect_interval

        # Active connections: device_id -> WebSocket
        self._connections: Dict[str, WebSocketClientProtocol] = {}

        # Connection state: device_id -> is_connected
        self._connected: Dict[str, bool] = {}

        # Authenticated device IDs
        self._authenticated: Set[str] = set()

        # Receive tasks: device_id -> asyncio.Task
        self._recv_tasks: Dict[str, asyncio.Task] = {}

        # Reconnect tasks: device_id -> asyncio.Task
        self._reconnect_tasks: Dict[str, asyncio.Task] = {}

        # Stop event
        self._stop_event = asyncio.Event()

        # This device's node ID
        self._node_id: str = ""

    def set_node_id(self, node_id: str) -> None:
        """Set this device's node ID for message origin field."""
        self._node_id = node_id

    async def connect(self, device_id: str, url: str) -> bool:
        """Connect to a WebSocket peer.

        Args:
            device_id: The device ID to associate with this connection.
            url: The WebSocket URL (e.g. "ws://192.168.1.100:8765").

        Returns:
            True if the connection was established (auth may still be pending).
        """
        # Cancel any existing reconnect task
        task = self._reconnect_tasks.pop(device_id, None)
        if task:
            task.cancel()

        try:
            ws = await websockets.connect(
                url,
                ping_interval=30.0,
                ping_timeout=10.0,
                max_size=64 * 1024 * 1024,
            )
            self._connections[device_id] = ws
            self._connected[device_id] = True
            logger.info("Connected to device: %s at %s", device_id, url)

            # Start receive loop
            self._recv_tasks[device_id] = asyncio.create_task(
                self._recv_loop(device_id, ws)
            )

            # Send hello to initiate authentication
            hello_msg = SyncMessage(
                type=MSG_TYPE_HELLO,
                origin=self._node_id,
                timestamp=self._now_iso(),
                payload={},
            )
            await ws.send(ProtocolCodec.encode_message(hello_msg))

            return True

        except Exception as e:
            logger.error("Failed to connect to %s at %s: %s", device_id, url, e)
            self._connected[device_id] = False
            return False

    async def disconnect(self, device_id: str) -> None:
        """Disconnect from a specific device.

        Args:
            device_id: The device ID to disconnect from.
        """
        # Cancel reconnect task
        task = self._reconnect_tasks.pop(device_id, None)
        if task:
            task.cancel()

        # Cancel receive task
        task = self._recv_tasks.pop(device_id, None)
        if task:
            task.cancel()

        # Close connection
        ws = self._connections.pop(device_id, None)
        if ws:
            try:
                await ws.close(1000, "Client disconnecting")
            except Exception:
                pass

        self._connected.pop(device_id, None)
        self._authenticated.discard(device_id)
        logger.info("Disconnected from device: %s", device_id)

    async def disconnect_all(self) -> None:
        """Disconnect from all devices."""
        device_ids = list(self._connections.keys())
        for device_id in device_ids:
            await self.disconnect(device_id)

    async def send(
        self,
        device_id: str,
        msg: SyncMessage,
        binary_payload: Optional[bytes] = None,
    ) -> bool:
        """Send a message to a specific device.

        Args:
            device_id: The target device ID.
            msg: The SyncMessage to send.
            binary_payload: Optional binary payload.

        Returns:
            True if the message was sent successfully.
        """
        ws = self._connections.get(device_id)
        if ws is None:
            return False

        try:
            encoded = ProtocolCodec.encode_message(msg, binary_payload)
            await ws.send(encoded)
            return True
        except Exception as e:
            logger.warning("Failed to send to %s: %s", device_id, e)
            return False

    def is_connected(self, device_id: str) -> bool:
        """Check if connected and authenticated with a device."""
        return device_id in self._authenticated

    def get_connected_device_ids(self) -> Set[str]:
        """Get the set of currently connected device IDs."""
        return set(self._authenticated)

    async def _recv_loop(self, device_id: str, ws: WebSocketClientProtocol) -> None:
        """Receive loop for a single WebSocket connection.

        Args:
            device_id: The device ID for this connection.
            ws: The WebSocket connection.
        """
        try:
            async for raw_message in ws:
                if isinstance(raw_message, str):
                    logger.warning("Received text frame from %s, ignoring", device_id)
                    continue

                data = bytes(raw_message)
                try:
                    msg, binary_payload = ProtocolCodec.decode_message(data)
                except Exception as e:
                    logger.warning("Failed to decode message from %s: %s", device_id, e)
                    continue

                await self._handle_message(device_id, msg, binary_payload)

        except websockets.exceptions.ConnectionClosed as e:
            logger.info("Connection closed for %s: code=%s", device_id, e.code)
        except asyncio.CancelledError:
            pass
        except Exception as e:
            logger.error("Error in recv loop for %s: %s", device_id, e)
        finally:
            self._connected[device_id] = False
            self._authenticated.discard(device_id)
            self._connections.pop(device_id, None)
            logger.info("Receive loop ended for device: %s", device_id)

    async def _handle_message(
        self,
        device_id: str,
        msg: SyncMessage,
        binary_payload: Optional[bytes],
    ) -> None:
        """Handle a received message from a connected device.

        Args:
            device_id: The sender device ID.
            msg: The decoded SyncMessage.
            binary_payload: Optional binary payload.
        """
        if msg.type == MSG_TYPE_AUTH_CHALLENGE:
            await self._handle_auth_challenge(device_id, msg)

        elif msg.type == MSG_TYPE_AUTH_OK:
            self._authenticated.add(device_id)
            logger.info("Authenticated with device: %s", device_id)

        elif msg.type == MSG_TYPE_AUTH_FAIL:
            logger.warning("Authentication failed with device: %s", device_id)
            await self.disconnect(device_id)

        elif msg.type == MSG_TYPE_PONG:
            # Heartbeat response, no action needed
            pass

        elif msg.type == MSG_TYPE_DEVICE_ONLINE:
            if self.on_device_online:
                self.on_device_online(msg.payload)

        elif msg.type == MSG_TYPE_DEVICE_OFFLINE:
            if self.on_device_offline:
                self.on_device_offline(msg.payload)

        elif msg.type in (MSG_TYPE_CLIPBOARD_SYNC, MSG_TYPE_CLIPBOARD_SYNC_BINARY):
            if self.on_clipboard_sync:
                item_type = msg.payload.get("item_type", "")
                item_data = msg.payload.get("item", {})
                self.on_clipboard_sync(device_id, item_type, item_data, binary_payload)

        else:
            logger.debug("Unhandled message type from %s: %s", device_id, msg.type)

    async def _handle_auth_challenge(self, device_id: str, msg: SyncMessage) -> None:
        """Handle an auth_challenge by computing HMAC and sending auth_response.

        Args:
            device_id: The device that sent the challenge.
            msg: The auth_challenge message.
        """
        nonce = msg.payload.get("nonce", "")
        if not nonce:
            return

        secret = self._secret_store.get(device_id, "")
        if not secret:
            logger.warning("No shared secret for device: %s", device_id)
            return

        hmac_value = compute_hmac(secret, nonce)
        response_msg = SyncMessage(
            type=MSG_TYPE_AUTH_RESPONSE,
            origin=self._node_id,
            timestamp=self._now_iso(),
            payload={"hmac": hmac_value},
        )
        await self.send(device_id, response_msg)
        logger.debug("Auth response sent for device: %s", device_id)

    async def _reconnect_loop(self, device_id: str, url: str) -> None:
        """Attempt to reconnect to a device with exponential backoff.

        Args:
            device_id: The device ID to reconnect to.
            url: The WebSocket URL.
        """
        max_attempts = 10
        for attempt in range(1, max_attempts + 1):
            if self._stop_event.is_set():
                break

            delay = self.reconnect_interval * (2 ** min(attempt - 1, 4))
            logger.info(
                "Reconnect attempt %d for %s in %.1fs",
                attempt, device_id, delay,
            )
            await asyncio.sleep(delay)

            if self._stop_event.is_set():
                break

            success = await self.connect(device_id, url)
            if success:
                logger.info("Reconnected to device: %s", device_id)
                break

        self._reconnect_tasks.pop(device_id, None)

    @staticmethod
    def _now_iso() -> str:
        """Return the current UTC time in ISO 8601 format."""
        from datetime import datetime, timezone
        return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")

File: android/app/src/main/java/com/integratedcaptureclipboard/android/sync/WebSocketClient.kt
package com.integratedcaptureclipboard.android.sync

import android.util.Log
import com.integratedcaptureclipboard.android.data.model.SyncMessage
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import okio.ByteString.Companion.toByteString
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

/**
 * WebSocket client manager for the ICC multi-device sync system.
 *
 * Manages multiple WebSocket connections to different devices using OkHttp.
 * Each connection is identified by a device ID and supports:
 * - Binary frame protocol (consistent with Python ProtocolCodec)
 * - Automatic ping/pong heartbeat (handled by OkHttp)
 * - Connection state tracking
 * - Send and broadcast operations
 */
class WebSocketClient(
    private val okHttpClient: OkHttpClient
) {
    companion object {
        private const val TAG = "WebSocketClient"
        private const val CONNECT_TIMEOUT_SECONDS = 10L
        private const val READ_TIMEOUT_SECONDS = 0L // No read timeout for persistent connections
        private const val PING_INTERVAL_SECONDS = 30L
    }

    /** Active WebSocket connections keyed by device ID. */
    private val activeConnections = ConcurrentHashMap<String, WebSocket>()

    /** Connection state listeners keyed by device ID. */
    private val connectionListeners = ConcurrentHashMap<String, WebSocketConnectionListener>()

    /** Connection state tracker: device ID -> is connected. */
    private val connectionStates = ConcurrentHashMap<String, Boolean>()

    /**
     * Connect to a WebSocket server at the given URL.
     *
     * @param deviceId The device ID to associate with this connection.
     * @param url The WebSocket URL (e.g. "ws://192.168.1.100:8765").
     * @param listener Callback for connection events and received messages.
     * @return The created WebSocket instance.
     */
    fun connect(deviceId: String, url: String, listener: WebSocketConnectionListener): WebSocket {
        // Close existing connection if any
        disconnect(deviceId)

        connectionListeners[deviceId] = listener

        val request = Request.Builder()
            .url(url)
            .build()

        val wsListener = object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: okhttp3.Response) {
                Log.d(TAG, "WebSocket opened for device: $deviceId")
                activeConnections[deviceId] = webSocket
                connectionStates[deviceId] = true
                listener.onConnected(deviceId)
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                // We use binary frames exclusively; text messages are unexpected
                Log.w(TAG, "Received unexpected text message from device: $deviceId")
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                try {
                    val (msg, binaryPayload) = SyncSerializer.decodeMessage(bytes.toByteArray())
                    listener.onMessage(deviceId, msg, binaryPayload)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to decode message from device: $deviceId", e)
                }
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closing for device: $deviceId, code: $code, reason: $reason")
                webSocket.close(1000, "Normal closure")
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closed for device: $deviceId, code: $code")
                activeConnections.remove(deviceId)
                connectionStates[deviceId] = false
                listener.onDisconnected(deviceId, code, reason)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: okhttp3.Response?) {
                Log.e(TAG, "WebSocket failure for device: $deviceId", t)
                activeConnections.remove(deviceId)
                connectionStates[deviceId] = false
                listener.onError(deviceId, t)
            }

        }

        val client = okHttpClient.newBuilder()
            .connectTimeout(CONNECT_TIMEOUT_SECONDS, TimeUnit.SECONDS)
            .readTimeout(READ_TIMEOUT_SECONDS, TimeUnit.SECONDS)
            .pingInterval(PING_INTERVAL_SECONDS, TimeUnit.SECONDS)
            .build()

        val ws = client.newWebSocket(request, wsListener)
        activeConnections[deviceId] = ws
        return ws
    }

    /**
     * Disconnect from a specific device.
     *
     * @param deviceId The device ID to disconnect from.
     */
    fun disconnect(deviceId: String) {
        val ws = activeConnections.remove(deviceId)
        if (ws != null) {
            ws.close(1000, "Client disconnecting")
            connectionStates[deviceId] = false
        }
        connectionListeners.remove(deviceId)
    }

    /**
     * Disconnect all active connections.
     */
    fun disconnectAll() {
        val deviceIds = activeConnections.keys.toList()
        for (deviceId in deviceIds) {
            disconnect(deviceId)
        }
    }

    /**
     * Send a SyncMessage (with optional binary payload) to a specific device.
     *
     * @param deviceId The target device ID.
     * @param msg The sync message to send.
     * @param binaryPayload Optional binary payload (e.g. image data).
     * @return True if the message was queued successfully.
     */
    fun send(deviceId: String, msg: SyncMessage, binaryPayload: ByteArray? = null): Boolean {
        val ws = activeConnections[deviceId]
        if (ws == null) {
            Log.w(TAG, "Cannot send to device $deviceId: not connected")
            return false
        }

        return try {
            val encoded = SyncSerializer.encodeMessage(msg, binaryPayload)
            ws.send(encoded.toByteString())
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send message to device: $deviceId", e)
            false
        }
    }

    /**
     * Broadcast a SyncMessage to all connected devices, optionally excluding some.
     *
     * @param msg The sync message to broadcast.
     * @param binaryPayload Optional binary payload.
     * @param exclude Device IDs to exclude from the broadcast.
     */
    fun broadcast(msg: SyncMessage, binaryPayload: ByteArray? = null, exclude: Set<String> = emptySet()) {
        val encoded = try {
            SyncSerializer.encodeMessage(msg, binaryPayload)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to encode broadcast message", e)
            return
        }

        val byteString = encoded.toByteString()
        for ((deviceId, ws) in activeConnections) {
            if (deviceId in exclude) continue
            try {
                ws.send(byteString)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to broadcast to device: $deviceId", e)
            }
        }
    }

    /**
     * Check if a device is currently connected.
     *
     * @param deviceId The device ID to check.
     * @return True if connected.
     */
    fun isConnected(deviceId: String): Boolean {
        return connectionStates[deviceId] == true
    }

    /**
     * Get the list of currently connected device IDs.
     *
     * @return Set of connected device IDs.
     */
    fun getConnectedDeviceIds(): Set<String> {
        return activeConnections.keys.filter { connectionStates[it] == true }.toSet()
    }

    /**
     * Callback interface for WebSocket connection events.
     */
    interface WebSocketConnectionListener {
        /**
         * Called when the WebSocket connection is established.
         *
         * @param deviceId The connected device's ID.
         */
        fun onConnected(deviceId: String)

        /**
         * Called when a binary message is received and decoded.
         *
         * @param deviceId The sender device's ID.
         * @param message The decoded SyncMessage.
         * @param binaryPayload Optional binary payload.
         */
        fun onMessage(deviceId: String, message: SyncMessage, binaryPayload: ByteArray?)

        /**
         * Called when the WebSocket connection is closed.
         *
         * @param deviceId The disconnected device's ID.
         * @param code The WebSocket close code.
         * @param reason The close reason string.
         */
        fun onDisconnected(deviceId: String, code: Int, reason: String)

        /**
         * Called when a WebSocket error occurs.
         *
         * @param deviceId The device ID with the error.
         * @param error The throwable error.
         */
        fun onError(deviceId: String, error: Throwable)
    }
}
